### PR TITLE
Draft training training script

### PR DESCRIPTION
**Pull request by yihjian**
Work done:
- Merged #38 
- Remove flags in model initialization to dnn/parameters.py
- Rename yolov3-tiny to yolov3_tiny to avoid import issues
- Rename yoloEncoder to yolo_encoder for consistency
- Fix loader to distinguish train and validation
- Mask out files with no labels in converter
- Add training script
- Add a debugging notebook with eager training
- Add a prediction demo script

Works need to be done:
- Finalize visualizer
- ~~Draft benchmark script~~
   - Need to vectorize
- Improve performance by moving some work from loss function to encoder
- Add support for transfer learning
- ~~Monitor with Tensorboard~~
- Update class names and file names for consistency

### CODE REVIEW

File: dnn/data/augmentation/detection.py [hunk 1]
```diff
@@ -221,6 +233,10 @@ def __call__(self, data, training=True):
         # apply transformation
         img_hw3 = tfa.image.transform(img_hw3, affine_coeffs, 
             interpolation='BILINEAR', output_shape=tf.cast(self.output_hw, tf.int32))
+        img_hw3 = tf.image.random_saturation(img_hw3, self.sat[0], self.sat[1])
+        img_hw3 = tf.image.random_hue(img_hw3, self.hue)
+        img_hw3 = tf.image.random_contrast(img_hw3, self.contrast[0], self.contrast[1])
+        img_hw3 = tf.image.random_brightness(img_hw3, self.brightness)
```
Comment (alvinsunyixiao) [inline dnn/data/augmentation/detection.py:239]: I think image-level augmentations should be moved to its own module, in parallel to `DetectionAugmentor`. May be create another `tf.Module` subclass in somewhere like `dnn/data/augmentation/image.py`?

**Review by alvinsunyixiao — COMMENTED** [review]
`yolo_encoder` looks pretty much like a prototype, we might eventually need to wrap it into a `tf.Module` or something similar

File: dnn/data/augmentation/detection.py (hunk 1 above)
Comment (yihjian) [reply dnn/data/augmentation/detection.py]: Done, moved into image.py

File: dnn/utils/yolo_encoder.py [hunk 2]
```diff
@@ -133,12 +123,12 @@ def transform_targets(y_train, anchors, anchor_masks, size):
...
 #Placeholder for lambda function
-'''
-Not sure why is it necessary to resize again
-'''
-def transform_images(x_train, size):
+def transform_images_train(x_train):
+    return x_train
+
+def transform_image_inference(x_train, size):
     x_train = tf.image.resize_with_pad(x_train, size, size)
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:130]: Why is this still needed?

File: dnn/utils/yolo_encoder.py [hunk 3]
```diff
@@ -133,12 +123,12 @@ def transform_targets(y_train, anchors, anchor_masks, size):
...
-    x_train = x_train / 255
+    x_train = x_train / 255 #it's possible that we don't need this line, left it there
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:131]: Let's not hardcode this normalization.

**Review by alvinsunyixiao — COMMENTED** [review]
See #40 for up-to-date TFRecord Format.

Can we package this into a `tf.Module` or just an object? We might want things to be tunable, say, the anchor sizes, whether to normalize input or not, etc. We can pass those parameters in from a constructor.

File: dnn/data/augmentation/detection.py [hunk 4]
```diff
@@ -2,6 +2,10 @@
 import tensorflow as tf
 import tensorflow_addons as tfa
 
+physical_devices = tf.config.experimental.list_physical_devices('GPU')
+for physical_device in physical_devices:
+    tf.config.experimental.set_memory_growth(physical_device, True)
+
```
Comment (alvinsunyixiao) [inline dnn/data/augmentation/detection.py:8]: I don't think this should be called in this file global scope. Whoever responsible for the top-level script should call memory management stuff. And there's a function `dnn.utils.mem.tf_set_memory_growth` for this.

File: dnn/utils/yolo_encoder.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        Args:
+            y_true: tensor of shape (boxes, (x1, y1, x2, y2, class, best_anchor))
+            grid_size: int, number of 'cut' per image. For yolov3-tiny, it's 13 and 26. 
+            masks: ordered anchor_ids of shape (1,n), e.g : (3,4,5)
+
+        '''
+        # mask out boxes that doesn't belong to this set of anchors
+        mask = tf.math.logical_and(
+            tf.math.greater_equal(y_true_n6[...,5], masks[0]), 
+            tf.math.less_equal(y_true_n6[...,5], masks[-1]))
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:46]: nit
but maybe a shorter way is to just use the operators `&` `>=` and `<=`

File: dnn/utils/yolo_encoder.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        Transform tensors that represent boxes into yolo_out
+
+        Args:
+            y_true: tensor of shape (boxes, (x1, y1, x2, y2, class, best_anchor))
+            grid_size: int, number of 'cut' per image. For yolov3-tiny, it's 13 and 26. 
+            masks: ordered anchor_ids of shape (1,n), e.g : (3,4,5)
+
+        '''
+        # mask out boxes that doesn't belong to this set of anchors
+        mask = tf.math.logical_and(
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:44]:
```suggestion
        mask_n = tf.math.logical_and(
```

File: dnn/utils/yolo_encoder.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            tf.math.greater_equal(y_true_n6[...,5], masks[0]), 
+            tf.math.less_equal(y_true_n6[...,5], masks[-1])) 
+        mask = tf.reshape(mask, (y_true_n6.shape[0],1))
+        mask = tf.broadcast_to(mask, y_true_n6.shape)
+        y_true_n6 = tf.boolean_mask(y_true_n6, mask)
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:49]:
```suggestion
        mask_n6 = tf.broadcast_to(mask_n[:, tf.newaxis], y_true_n6.shape)
        y_true_n6 = tf.boolean_mask(y_true_n6, mask_n6)
```
Is this tested? I don't think `y_true_n6` is of shape `(n, 6)`. If I am not mistaken, `tf.boolean_mask` returns a 1D array

File: dnn/utils/yolo_encoder.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                y_train_n6, grid_size, masks))
+            grid_size *= 2
+        
+        #y_out[0] = 13*13*3*6, y_out[1] = 26*26*3*6, y_out[2] = 52*52*3*6
+        # grid*grid*anchor*(x,y,w,h,confidence,class(2))
+        return tuple(y_outs)
+
+    #Placeholder for lambda function
+    def transform_images_train(self, x_train):
+        return x_train
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:133]: What is this for?

File: dnn/utils/yolo_encoder.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    def transform_image_inference(self, x_train, size):
+        x_train = tf.image.resize_with_pad(x_train, size, size)
+        x_train = x_train / 255 #it's possible that we don't need this line, left it there
+        return x_train
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:138]: I don't think this is still necessary given our augmentor can resize to a desirable shape?

**Review by alvinsunyixiao — COMMENTED** [review]
The variable names are still a little hard to follow as some of them did not have a shape suffix.

File: dnn/utils/yolo_encoder.py (hunk 9 above)
Comment (yihjian) [reply dnn/utils/yolo_encoder.py:138]: But during the inference pipeline, I don't the augmentor will be invoked

File: dnn/utils/yolo_encoder.py (hunk 8 above)
Comment (yihjian) [reply dnn/utils/yolo_encoder.py:133]: This is kind of a placeholder function for lambda x,y I guess, e.g.
``` python
train = loadDataset("whatever.tfrecords")
train.map(lambda: x, y : (
        dataset.transform_images(x), # Not sure if I can do just x
        dataset.transform_label(y)
))
```

File: dnn/utils/yolo_encoder.py (hunk 7 above)
Comment (yihjian) [reply dnn/utils/yolo_encoder.py:49]: It's tested before I did the wrapping up, I will double check

Comment (yihjian) [reply dnn/utils/yolo_encoder.py:49]: Probably I messed during renaming

File: dnn/utils/yolo_encoder.py (hunk 5 above)
Comment (yihjian) [reply dnn/utils/yolo_encoder.py:46]: oh didn't know that works

File: dnn/utils/yolo_encoder.py (hunk 7 above)
Comment (alvinsunyixiao) [reply dnn/utils/yolo_encoder.py:49]: Quick experiment
```python
mat = tf.constant([[0, 1], [2, 3]])
mask = tf.constant([[True, False], [True, False]])
# tf.boolean_mask(mat, mask) == [0, 2]
```
Here I think `mask` and `y_true_n6` are 2D arrays, the end result is 1D.

File: dnn/utils/dataLoader.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+
+FEATURES = {
+    'image': tf.io.FixedLenFeature([], tf.string),
+    'class_n': tf.io.FixedLenFeature([], tf.string),
+    'bbox_yxyx_n4': tf.io.FixedLenFeature([], tf.string),
+}
+
+
+class dataLoader(tf.Module):
```
Comment (alvinsunyixiao) [inline dnn/utils/dataLoader.py:14]: I don't think it's necessary to make it a `tf.Module`. But it does not hurt in any way.

File: dnn/utils/yolo_encoder.py (hunk 8 above)
Comment (alvinsunyixiao) [reply dnn/utils/yolo_encoder.py:133]: I see. That makes sense. Maybe it's more straightforward to make a function like `encode_dataset(self, raw_dataset)` and return a "mapped" dataset object. Otherwise, it might not seem obvious on the user side to use those helpers in such ways.

File: dnn/utils/yolo_encoder.py (hunk 9 above)
Comment (alvinsunyixiao) [reply dnn/utils/yolo_encoder.py:138]: I think encoders are for encoding labels for training purposes. During inference, it's mostly the others' responsibility to preprocess the images into desired shapes and feed them into the network, and so `yolo_encoder` should not be involved at all during inference.

File: dnn/utils/dataLoader.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        Takes tfrecord file, parse it, pass the data through augmentor, and return x_train, y_train
+        Args:
+            tfrecord: A tfrecord dataset
+        '''
+        example = tf.io.parse_single_example(tfrecord, self.feature_map)
+        augmentor_input = {
+            'image_hw3':    tf.io.decode_image(example['image'], channels=3),
+            'label_n':      tf.io.parse_tensor(example['class_n'], tf.int32),
+            'bbox_yxyx_n4': tf.io.parse_tensor(example['bbox_yxyx_n4'], tf.float32),
+        }
```
Comment (alvinsunyixiao) [inline dnn/utils/dataLoader.py:38]: Try print the following here.
```python
print(augmentor_input['image_hw3'].shape, augmentor_input['bbox_yxyx_n4'].shape)
```
If the shapes are `None`, you might need to do
```python
augmentor_input['image_hw3'].set_shape((None, None, 3))
augmentor_input['label_n'].set_shape((None,))
augmentor_input['bbox_yxyx_n4'].set_shape((None, 4))
```

File: dnn/train_tiny.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    # initialize model, loss function and optimizer
+    model = YoloV3Tiny(size, classes=num_cls, training=True)
+    loss = [YoloLoss(anchors[mask], classes=num_cls) 
+            for mask in np.array(anchor_masks, dtype=np.int8)] #cast to int 
+    optimizer = tf.keras.optimizers.Adam(lr = lr)
+    
+    ## Eager training, used for debugging, will remove in the future
+    avg_loss = tf.keras.metrics.Mean('loss', dtype=tf.float32)
+    avg_val_loss = tf.keras.metrics.Mean('val_loss', dtype=tf.float32)
+    for epoch in range(1, epoch + 1):
```
Comment (alvinsunyixiao) [inline dnn/train_tiny.py:54]: This eager training loop is better for debug, but slower than `model.fit`. We should test pieces of the training (network, tf.Dataset, losses) seperately in notebooks. And eventually, put together with `model.fit`.

Another way to be faster is by adding `tf.function` everywhere. But this is quite clunky in my opinion.

Comment (yihjian) [reply dnn/train_tiny.py:54]: I was trying to use eager mode to debug first so I commented out ```model.fit``` part.

**Comment by yihjian** [comment]
There is an ugly fix to resolve dataloader issues by changing tfrecord format to:
```
FEATURES = {
    'image': tf.io.FixedLenFeature([], tf.string),
    'class_n': tf.io.VarLenFeature(tf.int64),
    'bbox_yxyx_n4': tf.io.VarLenFeature(tf.float32),
}
```
when we read in data, we can do
```
augmentor_input = {
        'image_hw3':    tf.io.decode_image(example['image'], channels=3),
        'label_n':      tf.sparse.to_dense(example['class_n']),
        'bbox_yxyx_n4': tf.reshape(tf.sparse.to_dense(example['bbox_yxyx_n4']), (-1, 4)),
}
```
This can prevent wired iterating issues we reading dataset.

File: dnn/utils/yolo_encoder.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        '''
+        Takes a tf_record_dataset and return the encoded dataset
+        Args:
+            tf_record_dataset: a processed tf_record, i.e., the input
+                                shape should match and y should be in 0~1
+        '''
+        return tf_record_dataset.map(lambda x, y: (
+            self.transform_images_train(x),
+            self.transform_label(y)
+        ))
```
Comment (alvinsunyixiao) [inline dnn/utils/yolo_encoder.py:132]:
```suggestion
        return tf_record_dataset.map(lambda x, y: (
            self.transform_images_train(x),
            self.transform_label(y)
        ), num_parallel_calls=tf.data.AUTOTUNE)
```
Encoder is fairly expensive, try running it in parallel

**Review by alvinsunyixiao — COMMENTED** [review]
Try this if GPU becomes better utilized

File: dnn/utils/yolo_encoder.py (hunk 13 above)
Comment (yihjian) [reply dnn/utils/yolo_encoder.py]: doesn't seem to be very helpful

**Review by alvinsunyixiao — APPROVED** [review]
Training prototype working! Nice work!
